caramel/compiler/ocaml_to_erlang/fun.ml

```diff
@@ -261,7 +263,21 @@ and mk_expression exp ~var_names ~modules ~functions ~module_name =
                   (mk_expression arg ~var_names ~modules ~functions ~module_name))
           args
       in
-      Expr.apply name args
+      if wanted_arity > List.length args then
+        let rec gen_names acc n =
+          match n with
+          | 0 -> acc
+          | n ->
+              gen_names
+                (Name.var ("Caramelinternal" ^ string_of_int n) :: acc)
+                (n - 1)
+        in
+        let outer_names = gen_names [] (wanted_arity - List.length args) in
+        let lhs = List.map Pat.bind outer_names in
+        let idents = List.map Expr.ident outer_names in
+        let rhs = Expr.apply name (args @ idents) in
+        Expr.fun_ ~cases:[ FunDecl.case ~lhs ~guard:None ~rhs ]
```

Comment:
If I understand this correctly, you're

1. computed the arity of this particular function type, using `Uncurry.uncurry_tarrow/2`
2. checking if we're applying this function with the as many arguments as the arity indicated by the type from 1)
3. if we are missing arguments, create a function that will take the missing arguments and pass them on to the function from 1)

I think this will break for named and default arguments, so we need to have another look at this.

This can be problematic because fun's in Erlang can have a single arity, but multiple clauses. This means that a function `f ?(x = 1) y = x + y` would translate to 2 erlang functions:

```erlang
f(Y) -> f(1, Y).
f(X, Y) -> X + Y.
```

And a function with named arguments like `f ~x ~y = x / y` would need to be kept track until is _fully applied_ before we can call it.

For example, what does `let div_one = (f ~y: 1)` translate to?

```erlang
Div_one = fun (X) -> f(1, X) end,

% or
Div_one = fun (X) -> f(X, 1) end,
```

I think we may need to take a step back and rethink this. It's turning out to be a tad more complex than I thought id' be at first 😅

Reply:
> this will break for named and default arguments,

which are not supported right now anyway, but I understand we should think ahead...

Gleam supports labeled arguments, and I think those are probably fine to figure out at Caramel compilation time. Not sure about default arguments, they indeed seem tricky. What values are allowed? Just simple literals? If so then maybe they are just in the typedtree and could be fetched from there and printed when needed...

Reply:
> This means that a function `f ?(x = 1) y = x + y` would translate to 2 erlang functions:

I would propose only generate the full arguments version in the resulting Erlang, so only support optional / labeled arguments when called in from Caramel, not other langs.

> For example, what does `let div_one = (f ~y:1)` translate to?

This depends on the way f was declared, because of the order, and the rules of Ocaml erasure would need to be preserved. So assuming it is `f x ~y = x + y` which in erlang is just f(X, Y), then
```
Div_one = fun (X) -> f(X, 1) end,
```
and if `f ~y x = x + y` then `fun (X) -> f(1, X) end,`

Reply:
Or even generate the erlang functions with smaller arity for optional arguments, but only for erlang/elixir interoperability (don't use them from caramel because it's unnecessarily complicating things), and only if there is just one optional or have some rules like Elixir that they can be ommited inclusively starting from the end only.

In any case I think we can figure it out and if not, caramel is not 1.0 anyway :)